### DISCUSSION TITLE
Main approaches tried for DeepLabv3 Core ML conversion

### DISCUSSION THREAD

**Discussion by himanshunaidu** [TaskarCenterAtUW/iOSPointMapper#75] [General]
Hello All,
So since the day before I have been trying several approaches for converting DeepLabv3 Pytorch model to CoreML. 

I wanted to document all the important parts of this in a public page so that this can be discussed on in detail.

**Comment by himanshunaidu** [comment]
### Devices used for CoreML conversion

I ran this code in multiple environments, the main ones being:
1. MacBook Pro macOS Apple M2 Ventura 13.5.2
2. AWS EC2 Ubuntu instance Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz with NVIDIA Corporation TU104GL [Tesla T4]

As it turns out, based on my experiments, running the conversion on Mac worked better. Not ideal, but it isn't surprising.

**Comment by himanshunaidu** [comment]
### Approach 1: Reproduce the `from_deeplab.py` code from Ihor's `deeplabv3coreml` repository

Link: https://github.com/NaturalStupidlty/deeplabv3coreml/blob/main/from_deeplab.py

This file tries to use the DeepLabv3 model trained on CityScapes from the [DeepLabV3Plus-Pytorch](https://github.com/VainF/DeepLabV3Plus-Pytorch.git) repository


#### Result

While this successfully ran the conversion, the CoreML model performance ended up being significantly worse. 

Original Test Image:
<img src="https://github.com/user-attachments/assets/6c6f2626-98e3-4018-a17d-92749190a38a" width=300></img>

Pytorch Model Output: 
<img src="https://github.com/user-attachments/assets/5f9a4ef1-1224-40b2-bb58-bdfa983d554a" width=300></img>

CoreML Model Output:
<img src="https://github.com/user-attachments/assets/56b2143d-d39a-4d61-a946-b8fefebd5a76" width=300></img>

**Reply by himanshunaidu** [reply]
The output looks even worse when running this in the app on the camera live feed. It mostly looks like random noise. 

This issue happened in both environments.

**Comment by himanshunaidu** [comment]
### Approach 2: Reproduce the `from_torchvision.py` code from Ihor's `deeplabv3coreml` repository

Link: https://github.com/NaturalStupidlty/deeplabv3coreml/blob/main/from_torchvision.py

This file tries to use the DeepLabv3 models of different weights from the torchvision package

#### Result
While this successfully ran the conversion, the CoreML model performance ended up being significantly worse.

(I tried this for all the 3 variations of the deepLabv3 weights present in the `from_torchvision.py` file.)

Original Test Image:
<img src="https://github.com/user-attachments/assets/f87fbd8a-f160-4028-b85f-a7fcc10bef3d" width=200/>

PyTorch Model Output:
<img src="https://github.com/user-attachments/assets/999f8891-541e-4883-9368-8194776d3e3b" width=200/>

CoreML Model Output:
<img src="https://github.com/user-attachments/assets/633719c5-bca3-4585-a576-2cac814220bf" width=200/>

**Reply by himanshunaidu** [reply]
This issue happened in both environments.

**Comment by himanshunaidu** [comment]
### Approach 3: Use the models provided by Apple

Link: https://developer.apple.com/machine-learning/models/

Utilized the DeepLabV3.mlmodel file. 

I converted the MLMultiArray output to a UIImage using code from the following link:
https://dmytrohrebeniuk.medium.com/convert-mlmultiarray-to-image-for-pytorch-models-without-performance-lags-df93de9d2d46

Code used was similar to this: [coreml_convet_spec copy.txt](https://github.com/user-attachments/files/17597132/coreml_convet_spec.copy.txt)

#### Result
This is the only model that seems to have worked. 

Original Test Image:
<img src="https://github.com/user-attachments/assets/f87fbd8a-f160-4028-b85f-a7fcc10bef3d" width=200/>

CoreML Model Output:
<img src="https://github.com/user-attachments/assets/7ad7ab26-57fe-448c-8888-4724a4d61fbf" width=200/>

However, this doesn't solve the CoreML conversion problem we are having, since this approach involves using an Out-Of-Box model.

**Comment by himanshunaidu** [comment]
### Approach 4: Use the Basic version of official Apple documentation and build up from there

Tutorial Link:
https://apple.github.io/coremltools/docs-guides/source/convert-a-pytorch-segmentation-model.html

The default DeepLabv3 model from the tutorial works, however, we want a model that can be used as a VNCoreMLModel. The tutorial model accepts MLMultiArray, and we want a model that accepts Image. It would also be necessary for it to output an image, for performance as well as code convenience. 

It is possible to change the input format by changing the conversion logic. We must remember to normalize the image:
https://apple.github.io/coremltools/docs-guides/source/image-inputs.html

However, we would also like to change the output. When we change the output by taking into account the argmax requirement, we end up getting a failed compile, with the following error:
`RuntimeWarning: You will not be able to run predict() on this Core ML model. Underlying exception message was: Error compiling model: "compiler error:  Encountered an error while compiling a neural network model: validator error: main output 'output' must be a tensor of type Float32 or Float16".`

**Comment by himanshunaidu** [comment]
To note, there were dozens of other configurations tried, however, none of them seemed to have worked. 

The Apple documentation doesn't seem to have anything about this case, which is weird since they have clearly worked with conversion of DeepLabv3 model that accepts Image input. So it would have been prudent for them to document the process and make the documentation public.
